### DISCUSSION TITLE
assertSeeLiveWire - method not found

### DISCUSSION THREAD

**Discussion by finn-matti** [larastan/larastan#2129] [Q&A] [answered]
I have the following test code:

```php
$user = User::factory()->create();

$this->actingAs($user)
    ->get('bookmarks')
    ->assertSuccessful()
    ->assertSeeLivewire('pages.bookmarks');
```

Which gives the following error:

![CleanShot 2024-11-26 at 11 43 49@2x](https://github.com/user-attachments/assets/ec115852-f44e-40c4-acd4-701532df17a8)

Am I just doing something wrong? Or did I discover a bug?

I tried to search for similar issues, but I wasn't sure. So here we are.

I have larastan 3.0 installed and updated. I am running on PHP 8.3 and am using Laravel 11 and Livewire 3.5

**Comment by szepeviktor** [comment]
It may be an unknown macro
https://github.com/livewire/livewire/blob/047904314adc8a27046d042959a4c29c0feea6e2/src/Features/SupportTesting/SupportTesting.php#L80

**Reply by finn-matti** [reply]
This may be the case. How do I make it known though?

**Reply by szepeviktor** [reply]
Larastan may recognize that macro, try adding that file to `scanFiles:`
the other way is adding a stubs file.

**Reply by szepeviktor** [reply]
... and look at https://github.com/calebdw/larastan-livewire

**Reply by finn-matti** [reply]
I would've been surprised if scanFiles makes a difference. According to [the docs](https://phpstan.org/user-guide/discovering-symbols#third-party-code-outside-of-composer-dependencies):

>If your project uses some code that isn’t part of your Composer dependencies...

But that's not the case. Livewire is installed through Composer.

I anyways tried it, but it didn't get picked up.

```neon
parameters:
    scanFiles:
        - vendor/livewire/livewire/src/Features/SupportTesting/SupportTesting.php
```

>... and look at https://github.com/calebdw/larastan-livewire

This "only" supports, according to its README, computed properties.

I'll check out the stub file approach, but if you have other ideas, I'm all ears! Thanks!

**Reply by calebdw** [reply]
I generally **don't** have Larastan/PHPstan analyze my tests. They are tests and will fail during test time if something is wrong

**Reply by finn-matti** [reply]
Thanks for your perspective! However, since this is a learning project, I want to know how this works. So I tried to define stubs, which means I needed to create stubs for all referenced types in any given stub. For this case I came up with this (example):

```php
<?php

namespace Illuminate\Testing;

use Livewire\Component;

/**
 * Explicitly defined methods for macros.
 */
class TestResponse
{
    /**
     * Assert that a Livewire component is seen in the response.
     *
     * @param string|Component $component
     * @return $this
     */
    public function assertSeeLivewire(string|Component $component): self
    {
        return $this;
    }

    /**
     * Assert that a Livewire component is not seen in the response.
     *
     * @param string|Component $component
     * @return $this
     */
    public function assertDontSeeLivewire(string|Component $component): self
    {
        return $this;
    }
}

```

And I then have also created stubs for Component, ComponentRegistry, TestView and Assert, all the non-basic types referenced in the macros.

It doesn't make a difference.


I wonder two things:

1. According to the [docs](https://phpstan.org/user-guide/stub-files): "Stub files aren’t a replacement for [discovering symbols](https://phpstan.org/user-guide/discovering-symbols) so if you’re trying to fix errors like “Function not found” or “Class not found”, check out the [discovering symbols](https://phpstan.org/user-guide/discovering-symbols) guide instead." It doesn't say explicitly "method not found", but I assume that's included?
2. If methods can be added to Classes using stubs. Is this example stub I gave correct?

**Reply by calebdw** [reply, marked as answer]
The test macros are registered when the Livewire service provider is booted---however, the SupportTesting component hook early returns when the environment is not `testing`:


https://github.com/livewire/livewire/blob/047904314adc8a27046d042959a4c29c0feea6e2/src/Features/SupportTesting/SupportTesting.php#L12-L14

You likely are not running PHPStan in the `testing` environment so the custom macros are not being registered. Updating my `.env` to be `testing` allowed phpstan to discover the method

**Reply by finn-matti** [reply]
Wohoo! That was it. Somehow I didn't see that. And I guess it's no wonder, since there are no affordances to specify any env variables when using phpstan with PHPStorm. But this works.

@calebdw @szepeviktor Thanks to you both!

**Reply by calebdw** [reply]
I use Neovim (btw :sunglasses:), so not sure on PHPStorm specifics, but if you can customize the PHPStan command then might can update it to `APP_ENV=testing phpstan analyze`

**Reply by finn-matti** [reply]
For anybody else using PHPStorm a way that works is to provide a custom script that sets the env var. Something like this is enough:

```php
#!/usr/bin/env php
<?php

putenv('APP_ENV=testing');

// Get the command-line arguments passed to this script
$args = array_slice($argv, 1);

// Construct the PHPStan command
$phpstanCommand = 'vendor/bin/phpstan '.implode(' ', array_map('escapeshellarg', $args));

// Execute the command
passthru($phpstanCommand, $exitCode);

// Exit with the same code as PHPStan
exit($exitCode);
```

Make this thing executable:

```sh
chmod +x scripts/phpstan-with-testing.php
```

![CleanShot 2024-11-26 at 18 48 43@2x](https://github.com/user-attachments/assets/aff15685-b135-4871-9319-663df6f3953c)

Click there...

![CleanShot 2024-11-26 at 18 50 07@2x](https://github.com/user-attachments/assets/42417f29-6b98-4d52-bdef-e4999f2e30ee)

And then provide the script at the path. It should validate. If it does, you're good to go!

**Reply by Merdin** [reply]
> The test macros are registered when the Livewire service provider is booted---however, the SupportTesting component hook early returns when the environment is not `testing`:
> 
> https://github.com/livewire/livewire/blob/047904314adc8a27046d042959a4c29c0feea6e2/src/Features/SupportTesting/SupportTesting.php#L12-L14
> 
> You likely are not running PHPStan in the `testing` environment so the custom macros are not being registered. Updating my `.env` to be `testing` allowed phpstan to discover the method

Hi @calebdw ,

is it true that this still works the same?

**Reply by calebdw** [reply]
It should still work the same, but don't know about any updates on the Livewire side of things
